sshd-core/src/main/java/org/apache/sshd/common/session/helpers/AbstractSession.java

```diff
@@ -954,7 +954,9 @@ protected IoWriteFuture doWritePacket(Buffer buffer) throws IOException {
         synchronized (encodeLock) {
             Buffer packet = resolveOutputPacket(buffer);
             IoSession networkSession = getIoSession();
-            return networkSession.writePacket(packet);
+            IoWriteFuture future = networkSession.writePacket(packet);
+            buffer.rpos(buffer.wpos());
+            return future;
```

Comment:
The code is the following:
```
        synchronized (encodeLock) {
            Buffer packet = resolveOutputPacket(buffer);
            IoSession networkSession = getIoSession();
            IoWriteFuture future = networkSession.writePacket(packet);
            buffer.rpos(buffer.wpos());
            return future;
        }
```

The `buffer` is actually read from within the call to `resolveOutputPacket` and encoded into the `packet` variable.  So while the `packet` is yet to be written, the `buffer` won't be used anymore and is consumed at the time the `buffer.rpos(buffer.wpos())` is called.

Reply:
>>  So while the packet is yet to be written, the buffer won't be used anymore and is consumed at the time the buffer.rpos(buffer.wpos()) is called.

That was not what I meant - `networkSession#writePacket` might be asynchronous - i.e., posts the buffer for writing and returns immediately - so when it attempt to access `rpos` it might use the wrong one since we override it.

Reply:
Not sure to understand.  The `buffer` and `packet` variable are not the same objects.  I think we could also write:
```
  synchronized (encodeLock) {
            Buffer packet = resolveOutputPacket(buffer);
            buffer.rpos(buffer.wpos());
            IoSession networkSession = getIoSession();
            IoWriteFuture future = networkSession.writePacket(packet);
            return future;
        }
```
and it would also work.

Reply:
You are right - took me a bit to ensure that this is the case...

Reply:
Ok, so the problem is in the assumptions about the `writePacket`.  In the case of the nio2 implementation, it always wrap the data into a different buffer. Let me look a bit more into it.

Reply:
>> Ok, so the problem is in the assumptions about the writePacket.

I did not mean that there is any problem with the call since the packet is obviously a different instance than the original buffer. If there is some similar optimization that needs to be done on the packet then indeed we need to make it **after** being informed (via a listener) that the packet has been consumed.